sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala

```diff
@@ -50,10 +51,17 @@ object DynamicJoinSelection extends Rule[LogicalPlan] {
       mapStats.bytesByPartitionId.forall(_ <= maxShuffledHashJoinLocalMapThreshold)
   }
 
-  private def selectJoinStrategy(plan: LogicalPlan): Option[JoinStrategyHint] = plan match {
+  private def selectJoinStrategy(
+      plan: LogicalPlan,
+      joinType: JoinType): Option[JoinStrategyHint] = plan match {
     case LogicalQueryStage(_, stage: ShuffleQueryStageExec) if stage.isMaterialized
       && stage.mapStats.isDefined =>
-      val demoteBroadcastHash = shouldDemoteBroadcastHashJoin(stage.mapStats.get)
+      val demoteBroadcastHash = joinType match {
+        // doesn't make sense for outer joins since one side is preserved and join is not
+        // short circuited if the other side is empty
+        case LeftOuter | RightOuter | FullOuter => false
```

Comment:
does BHJ support `FullOuter` ? what about `LeftSemi` and `LeftAnti` ?

Reply:
BHJ can't support FullOuter - I put it in the list since it seems odd (though not wrong) to add `no_broadcast_hash` to a join that doesn't BHJ anyway.  Makes logs harder to read.
LeftSemi & LeftAnti could in principle support it - I don't know if they today.